From: Jeff Moyer <jmoyer@redhat.com>
Date: Tue, 8 Dec 2009 18:57:40 -0500
Subject: [block] cfq-iosched: improve hw_tag detection
Message-id: <1260298712-12756-8-git-send-email-jmoyer@redhat.com>
Patchwork-id: 21759
O-Subject: [RHEL6 PATCH 07/59] cfq-iosched: improve hw_tag detection
Bugzilla: 425895
RH-Acked-by: Vivek Goyal <vgoyal@redhat.com>

Related to bug 425895.

commit 1a1238a7dd48e48b3bba8f426a1d61c22c80d6d1
Author: Shaohua Li <shaohua.li@intel.com>
Date:   Tue Oct 27 08:46:23 2009 +0100

    cfq-iosched: improve hw_tag detection

    If active queue hasn't enough requests and idle window opens, cfq will not
    dispatch sufficient requests to hardware. In such situation, current code
    will zero hw_tag. But this is because cfq doesn't dispatch enough requests
    instead of hardware queue doesn't work. Don't zero hw_tag in such case.

Signed-off-by: Jeff Moyer <jmoyer@redhat.com>
Signed-off-by: Aristeu Rozanski <arozansk@redhat.com>

diff --git a/block/cfq-iosched.c b/block/cfq-iosched.c
index 68f43b2..1577b98 100644
--- a/block/cfq-iosched.c
+++ b/block/cfq-iosched.c
@@ -2257,6 +2257,8 @@ static void cfq_insert_request(struct request_queue *q, struct request *rq)
  */
 static void cfq_update_hw_tag(struct cfq_data *cfqd)
 {
+	struct cfq_queue *cfqq = cfqd->active_queue;
+
 	if (rq_in_driver(cfqd) > cfqd->rq_in_driver_peak)
 		cfqd->rq_in_driver_peak = rq_in_driver(cfqd);
 
@@ -2264,6 +2266,16 @@ static void cfq_update_hw_tag(struct cfq_data *cfqd)
 	    rq_in_driver(cfqd) <= CFQ_HW_QUEUE_MIN)
 		return;
 
+	/*
+	 * If active queue hasn't enough requests and can idle, cfq might not
+	 * dispatch sufficient requests to hardware. Don't zero hw_tag in this
+	 * case
+	 */
+	if (cfqq && cfq_cfqq_idle_window(cfqq) &&
+	    cfqq->dispatched + cfqq->queued[0] + cfqq->queued[1] <
+	    CFQ_HW_QUEUE_MIN && rq_in_driver(cfqd) < CFQ_HW_QUEUE_MIN)
+		return;
+
 	if (cfqd->hw_tag_samples++ < 50)
 		return;
 
